kolibri/core/utils/lock.py

```diff
@@ -0,0 +1,62 @@
+from contextlib import contextmanager
+from sqlite3 import OperationalError
+
+from django.db import connection
+from django.db import transaction
+
+
+class DummyOperation(object):
+    def __init__(self):
+        self.obj = None
+
+    def execute(self):
+        from kolibri.core.device.models import SQLiteLock
+
+        self.obj = SQLiteLock.objects.create()
+
+    def revert(self):
+        if self.obj:
+            self.obj.delete()
+
+
+class PostgresLock(object):
+    def __init__(self, key=None):
+        self.key = key
+
+    def execute(self):
+        query = "SELECT pg_advisory_lock({key}) AS lock;".format(key=self.key)
+        with connection.cursor() as c:
+            c.execute(query)
+
+    def revert(self):
+        query = "SELECT pg_advisory_unlock({key}) AS lock;".format(key=self.key)
+        with connection.cursor() as c:
+            c.execute(query)
+
+
+@contextmanager
+def db_lock():
+    lock_id = 1
+    if connection.vendor == "sqlite":
+        while True:
+            try:
+                with transaction.atomic():
+                    operation = DummyOperation()
+                    operation.execute()
+                    yield
+                    operation.revert()
+                break
+            except OperationalError as e:
+                if "database is locked" not in str(e):
+                    raise e
+    elif connection.vendor == "postgresql":
+        with transaction.atomic():
+            operation = PostgresLock(key=lock_id)
+            operation.execute()
+            yield
```

Comment:
I don't see `operation.revert()` here? That seems necessary because the type of advisory lock obtained is a session level lock, which should remain locked for the duration of the connection. I would suggest changing `pg_advisory_lock` to `pg_advisory_xact_lock` ([reference](https://www.postgresql.org/docs/9.1/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS)). Note that for the transaction level lock functions (`xact`), there aren't corresponding `unlock` functions because it would be cleared by COMMIT or ROLLBACK of the transaction, including if the connection is closed.

Reply:
Ah, thanks - that's my bad for misleading @apurva-modi here. I'll submit an update.